### ISSUE TITLE
Support new Smart Mapping Renderers

### ISSUE THREAD

**Issue by patrickarlt** [Esri/esri-leaflet-renderers#37]
http://blogs.esri.com/esri/arcgis/2015/03/02/introducing-smart-mapping/

I think what got introduced was a bunch of stuff for continuous ramps (size color ect...)

Here is a service that uses a continuous size ramp http://services.arcgis.com/rOo16HdIMeOBI4Mb/arcgis/rest/services/Aggregation%20of%20Trimet%20Transit%20Stops%20to%20Portland%20Neighborhoods/FeatureServer/0?f=pjson

**Comment by patrickarlt** [comment]
@kneemer this is looking good but I'm having some problems with opacity.

This service http://services.arcgis.com/rOo16HdIMeOBI4Mb/arcgis/rest/services/Aggregation%20of%20Trimet%20Transit%20Stops%20to%20Portland%20Neighborhoods/FeatureServer/0 renders as below.

![screen shot 2015-03-19 at 3 37 36 pm](https://cloud.githubusercontent.com/assets/378557/6742558/084fdcdc-ce4e-11e4-92d8-712b260487f1.png)

Backgrounds look fine but the opacity of the main symbols isn't rendered. I think you just need to hook the alpha value in the color to the opacity value in Leaflet and it should fix this.

**Comment by patrickarlt** [comment]
I'm also not entirely sure what is happening with this one http://services.arcgis.com/rOo16HdIMeOBI4Mb/arcgis/rest/services/Trimet%20Transit%20Stops%20In%20Portland%20Neighborhoods/FeatureServer/0

![screen shot 2015-03-19 at 3 36 53 pm](https://cloud.githubusercontent.com/assets/378557/6742616/5d13857a-ce4e-11e4-89d7-1542c3c9de19.png)

I created it by using the centered on theme in the map viewer. Looks like nothing is being passed through properly :(

![trimet_transit_stops_in_portland_neighborhoods](https://cloud.githubusercontent.com/assets/378557/6742659/c26fc7bc-ce4e-11e4-985f-91d88d4129c4.png)

![trimet_transit_stops_in_portland_neighborhoods](https://cloud.githubusercontent.com/assets/378557/6742657/bd38cff0-ce4e-11e4-8abc-25f32385db63.png)

**Comment by patrickarlt** [comment]
Same as the above but using the "Extreme" theme. Seems to be rendering the non-transparent stuff fine and nothing for everything else which is weird.

![screen shot 2015-03-19 at 3 59 17 pm](https://cloud.githubusercontent.com/assets/378557/6742919/2622120e-ce51-11e4-895c-12026e30b26f.png)

**Comment by kneemer** [comment]
The opacity of the circles in the first example you gave is the same at the circle level:

 * esri-leaflet-renderers
`<path stroke-linejoin="round" stroke-linecap="round" fill-rule="evenodd" stroke="rgb(255,255,255)" stroke-opacity="1" stroke-width="0.9999999999999749" fill="rgb(227,139,79)" fill-opacity="0.5137254901960784" d="M774,641.0000000000002A7.999999999999799,7.999999999999799,0,1,1,773.9,641.0000000000002 z" class="leaflet-clickable"></path>`

 * JS API
`<circle fill="rgb(227, 139, 79)" fill-opacity="0.5137254901960784" stroke="rgb(255, 255, 255)" stroke-opacity="1" stroke-width="1" stroke-linecap="butt" stroke-linejoin="miter" stroke-miterlimit="4" cx="242" cy="232" r="20.25" transform="matrix(1.00000000,0.00000000,0.00000000,1.00000000,0.00000000,0.00000000)" fill-rule="evenodd" stroke-dasharray="none" dojoGfxStrokeStyle="solid"></circle>`

However,  I appear to be adding more than one circle so that is most likely the issue there.

**Comment by kneemer** [comment]
@patrickarlt The other two issues should be resolved in commit adcf4e56a71fbdf2f0b255de9815f0db2c28caba

I made some last minute changes to make jslint happy and introduced a bug.  Could you check again, I was able to reproduce and the commit fixed my reproduction.

**Comment by patrickarlt** [comment]
@kneemer looking good.  At first glance both examples render perfectly but I think might have hit an edge case with http://services.arcgis.com/rOo16HdIMeOBI4Mb/arcgis/rest/services/Trimet_Stops_in_PDX_Neighborhoods/FeatureServer/0

![comparing_leaflet_to_arcgis_js_api](https://cloud.githubusercontent.com/assets/378557/6753409/ec2141b0-ced2-11e4-83c1-52ce018cff86.png)

It looks like the opacity is set differently just on this one shape.

Esri Leaflet

```html
<path stroke-linejoin="miter" stroke-linecap="butt" fill-rule="evenodd" stroke="rgb(153,153,153)" stroke-opacity="1" stroke-width="0.49999999999998745" fill="rgb(91,49,59)" fill-opacity="0.5019607843137255" class="leaflet-clickable" d="..."
></path>
```

JS API

```html
<path fill="rgb(255, 255, 255)" fill-opacity="0" stroke="rgb(153, 153, 153)" stroke-opacity="1" stroke-width="0.5" stroke-linecap="butt" stroke-linejoin="miter" stroke-miterlimit="4" path="...
"></path>
```

**Comment by patrickarlt** [comment]
@kneemer I figured out what is happening with the first example (circles). Because this layer is a polygon layer `createNewLayer` will be called multiple time because the geometry might be generalized and we want to update the geometry. The best way to do this is create a new layer and copy the latlngs over to the original layer. You can see this here https://github.com/Esri/esri-leaflet/blob/master/src/Layers/FeatureLayer/FeatureLayer.js#L66-L72

Unfortunately this means that your `createNewLayer` override https://github.com/Esri/esri-leaflet-renderers/blob/master/src/FeatureLayerHook.js#L68-L84 also gets called multiple times creating multiple circles. This might have always been a bug we just never noticed because we never made the circles transparent.

If you want to reproduce set https://github.com/Esri/esri-leaflet-renderers/blob/master/spec/comparisons.html#L173-L175 to 

```
lat: 45.52,
lon: -122.65,
zoomLevel: 11
```

and load the layer. The opacity if preserved but as you zoom in and out you can see additional circles being added. You probably just need to add a conditional so each circle only gets created once even if you see the layer id multiple times.

![comparing_leaflet_to_arcgis_js_api](https://cloud.githubusercontent.com/assets/378557/6754056/684824ee-ced7-11e4-97db-a361a888a1b2.png)

It also looks like there is an ordering issue, look at the top left.

**Comment by kneemer** [comment]
Commit 98febb4dad953d6f939561cd8ab377cd41fa0de9 fixes the duplicate symbols being drawn.

I'm not sure what the best approach is for the ordering issue.  Currently, the point layer is brought to the front right after the Feature Layer's onAdd method is called.  Maybe listening for the Feature Layer's load event would work?

**Comment by patrickarlt** [comment]
I'm not really sure how to tackle the ordering either but it still looks really good now anyway. Lets tackle the ordering issue later. All 3 examples now work! I'll see if i can break this some more :smile:

**Comment by kneemer** [comment]
@patrickarlt I added some tests for the visual variables sizeInfo and colorInfo.  I also tested the old layers at sample server 6.  Everything appears to be working normally.  I did find some discrepancy between the dashed lines in the JS API and Esri Leaflet Renderers.  However, this is not related to smart map changes and I will fix it later.  I have to work on some other things this week.  I'm going to do a release tomorrow morning if you and @jgravois cannot find any breaking changes.
